### ISSUE TITLE
[Rule Impl] Part 3 style_hover_persistent

### ISSUE THREAD

**Issue by drjoho** [IBMa/equal-access#647]
Follow on to #590 and #614

Write a rule based on detection of the :hover CSS pseudo class. To start, we will look at where :hover triggers another element to display and ensure the displayed element is the next element in the DOM from the triggering element.

- [x] analysis of the failure cases
- [x] discussion on which failure cases to cover since I found 6 additional cases.
- [x] resolution on failure cases of which there are 3
- [x] test cases written (Dev)
- [x] necessary files for the new rule created or edited as needed (Dev)
- [x] logic written in pseudocode (Dev)
- [ ] rule logic code (Dev) 
- [ ] test rule logic with test cases and add cases as needed (Dev)
- [ ] ensure messages are output and unit tests work (Dev)

Documentation at https://ibm.ent.box.com/file/521562780202

**Comment by drjoho** [comment]
Failure Case 3: Element hover persistent display of a sibling using the CSS "+" (This is referred to as an adjacent selector. The CSS adjacent selector will select only the element that is immediately preceded by the former element.)

HERE is a PASS case:
```
<!DOCTYPE html>
<html>
<head>
<style>
  div {
    background-color: yellow;
    display: none;
  }

  span:hover + div {
    display: block;
  }
  div:hover { display: block }
  </style>
</head>
<body>

  <span>Hover over me!</span>

  <div>I will show on hover</div>
  

</body>
</html>
```
This is a Failure case:
```
<!DOCTYPE html>
<html>
<head>
<style>
  div {
    background-color: yellow;
    display: none;
  }

  span:hover + div {
    display: block;
  }
  //div:hover { display: block }
  </style>
</head>
<body>

  <span>Hover over me!</span>
  <div>I will show on hover</div>
  

</body>
</html>
</html>
```

**Comment by drjoho** [comment]
Failure Case 4: Element hover persistent display of a sibling using CSS ~ (This is referred to as an general selector. This selector is similar to the ‘Adjacent Selector’ (+ sign), however, it is less strict. While an adjacent selector will only select the first element that is immediately preceded by the former selector, this one is more generalized. It will select any adjacent elements, as long as they follow the first element, as in the example below.)

HERE  is a PASS case:

```
<!DOCTYPE html>
<html>
<head>
<style>
  div {
    background-color: yellow;
    display: none;
  }

  span:hover ~ div {
    display: block;
  }
  div:hover { display: block }
  </style>
</head>
<body>

  <span>Hover over me!</span>

  <div>I will show on hover</div>

</body>
</html>
```
This is a failure case
```
<!DOCTYPE html>
<html>
<head>
<style>
  div {
    background-color: yellow;
    display: none;
  }

  span:hover ~ div {
    display: block;
  }
  div:hover { display: block }
  </style>
</head>
<body>

  <span>Hover over me!</span>

  <div>I will show on hover</div>
  <div>I will show on hover TOO</div>
  

</body>
</html>
```

**Comment by drjoho** [comment]
Failure Case 5: Hover display must be immediately below hover element, i.e., no space between the elements lower border and the child or sibling upper boarder.

Note: I am using Case 1 as the example here but the same will happen to cases 2-4.
Also I am using css margin to cause a break / space between two element containers.

Note: the position attribute can have five values: static (default), absolute, relative, fixed and sticky.

Here is a PASS case for  

```
<head>
<style>
  div {
    background-color: yellow;
    padding: 20px;
    display: none;
  }

  span:hover div {
    display: block;
  }
</style>
</head>
<body>
	<span>Hover over me!<div>I will show on hover</div></span>
</body>
</html>
``` 
Here is an example of a failure case

```
<!DOCTYPE html>
<head>
<style>
  div {
    margin: 1rem;
    background-color: yellow;
    padding: 20px;
    display: none;
  }

  span:hover div {
    display: block;
  }
</style>
</head>
<body>
	<span>Hover over me!<div>I will show on hover</div></span>
</body>
</html>
```

Here is another example of a failure (same case ?)
```
<!DOCTYPE html>
<head>
<style>
  div {
    background-color: yellow;
    padding: 20px;
    display: none;
  }

  span:hover div {
    display: block;
    margin: 1rem;
  }
</style>
</head>
<body>
	<span>Hover over me!<div>I will show on hover</div></span>
</body>
</html>
```
Note: I tried all values of the css attribute position with margin and none seem to make any difference.

**Comment by drjoho** [comment]
I just want to make sure we're comparing apples and apples here.

If _nothing_ appears on hover, that is not a failure of this SC. That is an example of a 'pass' (in the sense that it should not trigger a failure of On hover of Focus). I'm going to call that N/A to separate it from a pass or fail of something which meets the testing criteria.

So this is what I'm seeing:
Failure Case 1
Pass code - agreed
Fail code - disagree. N/A not fail (since no content appears on hover)

Failure case 2
Pass code - agreed
Fail code - disagree. N/A not fail (since no content appears on hover)

Failure case 3
Pass code - agreed
Fail code - **agreed (displayed content does not persist)**

Failure case 4
Pass code - agreed
Failure code - **agreed (since the SECOND line of the text displayed on hover is not persistent when moving point on it)**

Failure case 5
Pass code - agreed
Failure code - **agreed (text displayed on hover is not persistent when moving pointer on it** 

Failure case 6
Pass code - agreed
Failure code - disagree. N/A not fail (since no content appears on hover)

**Comment by drjoho** [comment]
```
        /**
         * Description: Trigger if on hover any content displayed is not persistent
         * Origin: Requirement 1.4.13
         */
        //
        // Trigger logic:
        //
        // First, logic for the sibling failures
        // 
        // 1. Do we have an element with a hover selector,
        //    e.g., a:hover, span:hover, etc...?
        //
        // 2. is the hover followed by either the + or ~ css combinators?
        //
        //    e.g., :hover+ or :hover~ (after triming spaces)
        //
        // 3. is the element after the css combinator + or ~ followed by an element 
        //    with a display attribute with any value but none? e.g, 
        //
        //    [element1]:hover + [element2] {
        //       display: block;
        //
        //    [element1]:hover ~ [element2] {
        //       display: block;
        //
        // 4. if we found + and there is one sibling directly after [element1] in the body
        //    of type [element2]. 
        //    if 1-4 are true Trigger Failure 1
        //
        // 5. if we found ~ and there are two or more siblings directly after [element1] 
        //    in the body of type [element2] - note there cannot be other elements inbetween 
        //    the siblings. if 1-3 and 5 are true Trigger Failure 2
        //
        //
        // Logic for margin failure (or more generally any space between the bottom of 
        // [element1] and the top of [element2])
        //
        // 1. Do we have an element with a hover selector,
        //    e.g., a:hover, span:hover, etc...?
        //
        // 2. is the hover followed by any of the four css combinators?
        //
        //    descendant selector (space)
        //    child selector (>)
        //    adjacent sibling selector (+)
        //    general sibling selector (~)
        //
        //    e.g., [element1]:hover [element2], or [element1]:hover > [element2],
        //          [element1]:hover + [element2], [element1]:hover ~ [element2]
        //          Note: for the one with a space we cannot trim the space, the
        //                others may have spaces trimmed
        //
        // 3. does the element after [elemnent1]:hover contain one of the four
        //    css combinators followed by [element2] 
        //    with a display attribute with any value but none?? e.g, 
        //
        //    [element1]:hover [element2] {
        //       display: block;
        //
        //    [element1]:hover > [element2] {
        //       display: block;
        //
        //    [element1]:hover + [element2] {
        //       display: block;
        //
        //    [element1]:hover ~ [element2] {
        //       display: block;
        //
        // 4. if we found space for the css combinator [element1] must have one descendent
        //    of type [element2] in the body. 
        //
        // 5. if we found > for the css combinator [element1] must have one child
        //    of type [element2] in the body. 
        
        // 7. if we found + is there at one sibling directly after [element1] of type
        //    [element2] in the body. 
        //
        // 8. if we found ~ is there at two or more siblings directly after [element1] of type
        //    [element2] in the body - note there cannot be other elements inbetween the siblings. 
        //
        // 9. if either [element 1] or [element 2], contains a margin attribute with a positive value
        //    and 1-3 is true and one 1-4 is true then Trigger Failure 3.
```

**Comment by drjoho** [comment]
Working on adding a few more test cases that will determine what if anything needs to change in the above logic if multiple style hover persistent issues are detected both in the css and the body of the document.

**Comment by drjoho** [comment]
Most of rule logic code written will finish in Part 4
